### ISSUE TITLE
Git branching model

### ISSUE THREAD

**Issue by jensscherbl** [symphonycms/symphony-2#2268]
Recently stumbled across this great article and thought this would be a great fit for Symphony as well.

http://nvie.com/posts/a-successful-git-branching-model/

Just wanted to put it up for discussion here.

**Comment by brendo** [comment]
Alrighty. What the hell are we going to do here.

*Multiple branches*
My thoughts are we don't the resources to manage multiple branches or releases, so creating all these additional branches is giving a false impression, and could make coordinating new releases tricky. Not merging a pull request for a year is almost as bad as having the code sit stagnant for a year.

*Timed releases*
I'm generally not a super big fan of timed releases in general. The majority of my Symphony work has been with agencies, so my thinking is a little skewed, but I think having a release each month would frighten developers. By the time a project is finished, it could already be behind by 2-3 versions!

Coordinating a release is probably 2-3 hours work at the minute, so it's not too bad overhead wise, but our development is fairly sporadic, with minimal commits one month and then a heap in the next month. If we go down the road of skipping releases, it'll happen more likely than not, and at that point, what's the point of the monthly release :)

----

What about this setup, which is pretty true to the OP:

- `master` The most recent stable build of Symphony
- `integration` Our bleeding edge development
- `hotfix` Synced with `master` and `integration`, commits here will be pushed to `master` to form minor releases

When we think we've got a happy place for doing a release, we can spin off a `[version number]` branch so we can start to tie down a release with just bug fixes.

Our first release branch is likely to be 3.0.0 because the open pull requests mean that a 2.7.x is impossible (PHP 5.4+, Cookies & Sessions totally replaced and Database abstracted to PDO).

**Comment by jensscherbl** [comment]
**tl;dr**

+1

> Multiple branches

The development model I have in mind is like this...

While adding new features and making bigger changes for a hypothetical `2.7.0` release on a `2.7.0`-branch, a `2.6.x`-branch could exist to accumulate and release all the smaller fixes at a faster pace.

After `2.7.0` is released and work on a `2.8.0`-branch has started, `2.6.x` would be replaced by `2.7.x`.

The `2.8.0`-branch would even be created as soon as `2.7.0` becomes release candidate, to avoid last minute feature creep and irresponsible code changes at this stage.

A `3.0.0` branch would be an edge case and occur less often. It would only exist to accumulate bold moves and breaking changes.

That way, developers currently working on a `2.6.x` project could hold off on updating a little longer (until `2.7.x` reaches greater stability with `2.7.1` or `2.7.2`) while still getting most of the more recent bugfixes. Using `integration` in production to get critical bugfixes should be a thing from the past.

> What about this setup
> - `master` The most recent stable build of Symphony
> - `integration` Our bleeding edge development
> - `hotfix` Synced with master and integration, commits here will be pushed to master to form minor releases
> When we think we've got a happy place for doing a release, we can spin off a [version number] branch so we can start to tie down a release with just bug fixes.

I guess we can actually achieve the development model outlined above with the setup you suggested (our current setup + `hotfix`-branch + release-branches).

> Not merging a pull request for a year is almost as bad as having the code sit stagnant for a year.

Making better use of [milestones][2] and tagging issues and pull requests with version numbers for upcoming releases could help making things more transparent to contributors.

[1]: https://github.com/symphonycms/symphony-2/issues/2268#issuecomment-64291701
[2]: https://github.com/symphonycms/symphony-2/milestones

> **Timed releases**
> Development is fairly sporadic, with minimal commits one month and then a heap in the next month. If we go down the road of skipping releases, it'll happen more likely than not, and at that point, what's the point of the monthly release :)

I agree regarding major and feature releases. Would still make sense for hotfix releases.

> Having a release each month would frighten developers. By the time a project is finished, it could already be behind by 2-3 versions!

Shouldn't be a problem now that we use semantic versioning? I guess having a painless update with bugfixes every month can be a little scary in the beginning, but developers will get used to it... ;)

> Our first release branch is likely to be 3.0.0 because the open pull requests mean that a 2.7.x is impossible (PHP 5.4+, Cookies & Sessions totally replaced and Database abstracted to PDO).

Ok then.

**Comment by nitriques** [comment]
I do agree with most of the things, but I still feel like having version'd branches is the way to go.

a hotfix branch does not tells which version is currently hotfix'd.

**Comment by jensscherbl** [comment]
> a hotfix branch does not tells which version is currently hotfix'd.

Current version (`master`) only. Plan is to release bugfixes for the current version while developing the next major or minor release.

**Comment by nitriques** [comment]
> Plan is to release bugfixes for the current version while developing the next major or minor release.

So you'll release a new version each time there is a bug fix ? I would wait at least one week before doing so, in case we can have multiple ones.... am I wrong ?

**Comment by jensscherbl** [comment]
> So you'll release a new version each time there is a bug fix ?

No, just on a more regular basis, like once a month or so.

**Comment by nitriques** [comment]
> No, just on a more regular basis, like once a month or so.

Ok. So I send a PR on hotfix, it then gets merges into master on a rebase and then we have to rebase integration to get latest master fixes, am I right ?

**Comment by jensscherbl** [comment]
> So I send a PR on hotfix, it then gets merges into master on a rebase and then we have to rebase integration to get latest master fixes, am I right ?

I guess so.

**Comment by nitriques** [comment]
I'll close this issue now since we have a model, as described here: https://medium.getsymphony.com/announcing-symphony-cms-2-x-lts-release-13d5f57b2370

`2.7.x`, `lts` and `master` will NEVER get rebased, as this is prone to conflicts.

All PR must be submitted against a versioned branch. Future version branch can be rebased from time to time.
